sympy/tensor/indexed.py

```diff
@@ -359,7 +259,10 @@ def args(self):
             return self._args
 
     def _hashable_content(self):
-        return Expr._hashable_content(self) + (self._shape,)
+        if self._shape:
+            return super(IndexedBase, self)._hashable_content() + (self._shape,)
+        else:
+            return super(IndexedBase, self)._hashable_content()
```

Comment:
The `if` is complicating the code.
Unless benchmarks show significant savings, it should be removed.
(Edsger Dijkstra: "Premature optimization is the root of all evil." The issue is that if you optimize, the code base becomes more complicated, and hence more prone to error; even if each individual optimization does not add significant complication, the code base as a whole becomes more and more fragile, has more code paths to test, and more code path combinations that might need testing due to unexpected interactions. Also, maintainers have to make the unfortunate choice between slower reading and skipping optimization details which might have been relevant.)